main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -114,35 +114,36 @@ object Kinder {
    */
   private def visitStruct(struct0: ResolvedAst.Declaration.Struct, taenv: Map[Symbol.TypeAliasSym, KindedAst.TypeAlias], root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Struct, KindError] = struct0 match {
     case ResolvedAst.Declaration.Struct(doc, ann, mod, sym, tparams0, fields0, loc) =>
-      // if not annotated tparams are assumed to be Star except the last one which is Eff
-      val tparams1 = tparams0 match {
+      // In the case in which the user doesn't supply any type params,
+      // the parser will have already notified the user of this error
+      // The recovery step here is to simply add a single type param that is never used
+      val tparams1 = if (tparams0.tparams.isEmpty) {
+        val regionTparam = ResolvedAst.TypeParam.Unkinded(Name.Ident("$rc", loc), Symbol.freshUnkindedTypeVarSym(Ast.VarText.Absent, isRegion = false, loc), loc)
+        ResolvedAst.TypeParams.Unkinded(List(regionTparam))
+      } else {
+        tparams0
+      }
+      // if not annotated, tparams are assumed to be Star except the last one which is Eff
+      val tparams2 = tparams1 match {
```

Comment:
This technique still seems a little off. But we're in uncharted waters with the restriction on the last parameter.

I propose the following: We ignore whether the tparams are kinded or not. I.e., we just do `tparamsN = tparams0.tparams` so we just have a list of TypeParams. Then you can add the last one if needed.

Then we map over them with `getKindEnvFromTypeParamDefaultStar` and `KindEnv.merge` the result.

Finally we can map over the params with `visitTypeParam` and then deal with fields.

Reply:
This won't work because then the last tparam will initially be given kind `Star` and then with the merge we try to assign it kind `Eff`, the merge will fail with a `Validation.hardFailure`(see the `+` operator of `KindEnv`). I initially tried this, but it didn't work, hence my current method of kinding before inserting I merged a while ago.